### ISSUE TITLE
Monster list is getting unmanageable

### ISSUE THREAD

**Issue by burnhamrobertp** [AdventureLookup/AdventureLookup#236]
Do we have an official policy on what monsters are allowed into the "common monsters" list and what monsters are allowed into the "boss monsters" list?

Do non-standard (i.e. not in kobold fight club with "all sources" checked) monsters get put into either? If so, what value does that provide since they're overly specific?

For Boss monsters, such as "The Emerald Queen" which is like "Strahd" in a sense, but is significantly less well known. Is the official policy to put "Nymph" (or whatever the Emerald queen is) and "Vampire" respectively and have the description mention the name so that anyone who searches for "The Emerald Queen" still finds the adventure?

I tried to get in and clean up some of the monsters and it is entirely untenable. The list of monsters is insanely long and there are so many "1 use" and "2 use" monsters that I'm twitching. There are also several change requests open along the lines of this very same thing.

@JohnnyFlash or @MattColville I'm looking to you guys for your opinions in particular.

For the $0.02 its worth, I'm not a fan of the overly specific boss monsters or common monsters. I know that I want to find adventures with "Slime" or "Troll" or "Vampire" and oftentimes, even when "Undead Goblin" would be really nice, all I'm actually looking for is "Undead."

**Comment by NorNamlin** [comment]
Details, for Boss and Common monsters are to be listed per Matts instructions here:

https://www.reddit.com/r/AdventureLookup/comments/6tjaug/matts_guidelines_for_entries_fieldbyfield/

**Comment by burnhamrobertp** [comment]
@NorNamlin I've reviewed that in days past. The opening sentence is what I'm looking for clarification on. The overwhelming majority of monsters that have <4 uses don't seem to follow the common sense of what people are likely to be searching for. There are some that are good and reasonable, Angel for example, but for every good one there seemed to be another 20 that were just uselessly specific and/or custom monsters.

**Comment by NorNamlin** [comment]
Perhaps when a new submission is entered, it should not automatically be added to the data base until reviewed by a curator for basic submission requirements.  Such as monster format and other needed information.

This would eliminate the "1 use" and "2 use" issues you are seeing and allow them to be changed to an agreed upon format.  Goblin vs goblin(s) would certainly be one example of this among others.

**Comment by burnhamrobertp** [comment]
I think "all new submissions must be agreed upon" approach has come up in past discussions and identified as less than desirable. If every entry is low quality, that's a UX issue with AL that we need to fix in the code (either more helpful, better guidelines, or limiting submissions from repeat offenders. If "low quality" is rare, then it doesn't make sense that all submissions are punished for the sake of the worst of the worst.

**Comment by MattColville** [comment]
Ok help me understand what we mean by "unmanageable." Because it sounds like the question is "how do we best want to categorize these things" but it's not clear to me how that's related to managing the monster list. What managing are we expecting people to do?

Robert, when you say "Go in and clean up" give me some examples of what you're trying to fix. 

There's a lotta monsters in D&D and one of the things Adventure Lookup is good for is not only searching, but then telling. You search for an adventure, then you look at the entry and learn a LOT about that adventure! What you learn, looking at the entry, is just as useful as what you're searching for. 

I definitely think the Boss Monster field should have each boss listed by their proper name and type, with type being the most important of the two.

**Comment by burnhamrobertp** [comment]
If you go to search using the "common monster" filter - for example, and the monster(s) you're looking for aren't in the top 5, you need to click "more". There are almost 2,000 "common monsters"  in AL right now. As soon as you click on "more" the filters list becomes uselessly long. It is barely tolerable (having used it a few times for adventures with trolls, for example) if you know exactly what you're looking for. 

The current implementation isn't even as bad as it could be; right now clicking on "more" only loads 1,000 monsters. 

If we look at some numbers, the badge number next to each common monster is the number of adventures that monster is referenced in. The top (by number of uses) 1000 common monsters have a total adventure-reference count of 7656. The top 250 of common monsters have a total adventure-reference count of 5671. Why are these numbers meaningful? See below.

> Robert, when you say "Go in and clean up" give me some examples of what you're trying to fix.

- Duplicate monsters via typo
- Duplicate monsters via plurality
- Overly specific monsters (human warrior, human rogue, etc.)
- Monsters with no uses at all (presumably because they were entered mistakenly or because they were entered but then were cleaned up by a curator at some point)

I spent 3 hours doing this and managed to get to the middle of the "C" monsters before I was frustrated and at the end of my ropes. I had meant to get on and do development for AL, not mindless data cleanup...but it was data cleanup that needed to occur; the bit of research I had done for "trolls" the week prior had turned up some nasty-formatted adventures. If this represented the state of the data at large (and I had no reason to suspect it didn't), then the data in AL was in a pretty poor state.

At the end of the 3 hour session, I happened to notice something. The common monster "Goblin" didn't have any "Types" (Humanoid). Last night I spent another hour and a half looking up all the common monsters I could think of. Hobgoblin, Troll, Ogre, Giant Spider, Orc, etc. and none of them had a "type." Not a single one. These were the bread and butter of monsters but were incomplete.

This goes back to my earlier point; there is literally so much data in AL (we're only talking about "Common Monsters" let alone Boss Monsters, Notable Items, description quality, or any of the additional features that we want to add) and almost all of it is useless. The 250 most frequently used monsters (of 2000) constitute 75% of all references. But of those 250, how many actually have a type and are up to date? How many don't have duplicates in the form of plurality mistakes that might be interfering with usability?

![image](https://user-images.githubusercontent.com/231052/31210303-4217a2f8-a95f-11e7-965c-f5a139dc7144.png)
![image](https://user-images.githubusercontent.com/231052/31210330-6e9c2c2c-a95f-11e7-978b-65d78edeb8bc.png)
![image](https://user-images.githubusercontent.com/231052/31210345-7ffae058-a95f-11e7-8396-786cc7eac782.png)

If the data in AL is a 100 foot high mountain of tangled spaghetti, figuratively speaking, how can I reasonably look at it and "serve" it to users? How can anyone actually contribute to the project if the data is in this sad state of affairs? How do you have developers handle edge cases where there are over 1400 "common monsters" that have 1 _or even 0_ uses?

**Comment by burnhamrobertp** [comment]
The problem I'm seeing here is scalability. The current system is already painful. Its clear no one has gone through the common monsters, boss monsters, notable items, etc. to "clean up" (this has to be an admin because merely "merging" stuff as a curator doesn't remove it from the database). And AL hasn't really been in use all that long.

**Comment by calculuschild** [comment]
I agree, a giant list of monsters thousands of entries long is not useful in its current state.

In issue #201 I proposed earlier a sort of nested sorting approach to help classify and sub-classify all the monsters. This would ease searching for a particular monster type (i.e., I want zombies. I don't care if they are "goblin zombies", "human zombies", or "aquatic zombies") as well as reveal similar monster types to DMs to help expand their knowledge (let's see... zombies... oooh, there's a DRAGON ZOMBIE? How cool is that! Let's do that adventure!)

I think that way DMs could easily sort by _type_ of monster and drill down to a specific monster that meets their needs, even if it happens to be a single-use monster. Or, they could just select one of the overarching types and now you have a list of adventures including any type of zombie at all, and not worry that the only adventures they are going to see are the normal garden-variety zombies, and miss out on all those cool dragon zombies.

* Undead v
  * Vampires >
  * Zombies v
    * Abomination Zombies >
    * Humanoid Zombies >
    * Beast Zombies v
      * Zombie Bear
      * Zombie Fox
      * Zombie Fish
      * **Zombie Dragon**
  * Spirits >
  * ...
* Devils / Demons >
* Beasts >
* Humanoids >
* ...

Plus, how cool would that be to have a taxonomy of all dungeons and dragons monsters? Kobold Fight Club has a little bit of this, but I think we could refine it.

**Comment by gn000we** [comment]
@burnhamrobertp 
Is there a way to extract all the prod monsters and categories, so i can load them into my dev db?

**Comment by burnhamrobertp** [comment]
No, unfortunately not with how everything is organized presently. If you use the fixture specified in the readme, though, you do get monster data that you can use to test with. if you wanted to expand that you could always locate the fixture that has that data and add more to it!

**Comment by gn000we** [comment]
@burnhamrobertp  could you then give a rough thought process or steps you were taking to clean the data? Were the issues user input variance and linking/associating? Was it the structual consequences and relationships within the db?

**Comment by burnhamrobertp** [comment]
Typographical variance is one problem, but "usefully distinct" is probably a better description of my biggest concern. Unique and "unique variant" monsters constitute a meaningful amount of data in the monsters tables (common and boss). Unique-ish monsters aren't going to be used by anyone. 

![image](https://user-images.githubusercontent.com/231052/32913380-ba9d5652-cad7-11e7-89db-91420d7d1057.png)

Who is going to search for "Circe"? If we go up a few entries, we see this problem isn't small.

![image](https://user-images.githubusercontent.com/231052/32913429-dfc2c9da-cad7-11e7-8b51-6db29784f080.png)

Who is going to search for Chariot Steed, Charon, Chevall, Child of the Briar, or Children of Maalpherus? There are thousands of these creatures that have (and will only ever have, in all likelihood) only one reference in AL.

**Comment by gn000we** [comment]
I find I battle with this myself in RL! Categorisation to handle large issues, but when I see the details, it gets defined into lots of inividual things.

**Comment by gn000we** [comment]
@burnhamrobertp  would a human voting system on here be useful? Data is data. Technically we shoukd be able to marry this up.

**Comment by burnhamrobertp** [comment]
We have curators who clean it up, but its happening on a scope that exceeds human-management's ability to keep it in check, realistically.

**Comment by burnhamrobertp** [comment]
But as you can tell from this issue, there is contention that a problem even exists, so code changes pertaining to this issue are unlikely to be approved and merged in until there is a concensus about the presence (or absence) of a problem in this regard.

**Comment by gn000we** [comment]
yes, i agree. is there a form of updates/scripting used once identified?

**Comment by gn000we** [comment]
people saying it's not a problem: yeh i can see that, but tagging instead of categorisation should suffice. Tag freely, but categorise strictly (chaos & lawful satisfied).

**Comment by burnhamrobertp** [comment]
> is there a set of updates/scripting used once identified?

How do you mean?

> name: Ent #"brair", #"Child of brair", #"zombie brair"

Not sure what this is intended to tell me, as far as your example is intending.

Categorization isn't quite that simple though, as your example, as simple as it may be, indicates. A Zombie Briar isn't a plant anymore, even if it did originate (lore-wise) from a plant. There are lots of monsters that could / would be in multiple categories, and - worse yet - many monsters with the same name move between different categories in different editions of D&D.

Finally, I ask you, even if a developer builds this categorization functionality and anyone submitting an event was required to provide a category for each monster provided (and presuming that they did)...does that actually solve the problem I brought up above?

> Who is going to search for Chariot Steed, Charon, Chevall, Child of the Briar, or Children of Maalpherus? There are thousands of these creatures that have (and will only ever have, in all likelihood) only one reference in AL.

That the monsters are now in categories doesn't change the fact that they won't be used by name anymoreso than they are now.

**Comment by gn000we** [comment]
> 
    Duplicate monsters via typo
    Duplicate monsters via plurality
    Overly specific monsters (human warrior, human rogue, etc.)
    Monsters with no uses at all (presumably because they were entered mistakenly or because they were entered but then were cleaned up by a curator at some point)

Are the above still a problem?

**Comment by gn000we** [comment]
It's ok, i'm working on the scripts and a potential solution (for fun)

**Comment by burnhamrobertp** [comment]
You failed to address the issues I raised in the last message I posted.

**Comment by calculuschild** [comment]
Apologies in advance. This got a lot longer than I intended.

I think it is pretty clear that some categorization is needed, and probably a nested tree structure. Even if the list of monsters is whittled down to the top 250, that's still a huge list of monsters that are probably not all familiar to every DM. If this website is to help a DM find an adventure with a particular type of monster, I think letting him search first *by type* and then branch down to a particular subtype will be much more useful than trying to remember the particular name of a fictional creature. If I've never heard of an Otyugh, I would never know to find it, but if I know I want some kind of Aberration, I can at least start there.

> There are lots of monsters that could / would be in multiple categories

I don't believe there is any issue with a monster being part of multiple categories. It even makes a lot of sense and could be very useful to new DMs, and storing this in a database isn't impossible. One-to-many database relationships are a common structure.

> Who is going to search for "Circe"? If we go up a few entries, we see this problem isn't small.

I agree with this sentiment for the most part. I don't think we necessarily need "Super guardian goblin zombie of the dark" when we could just put "goblin zombie". But at the same time, *who is going to search for "Otyugh"?* That one is a legitimate Monster Manual creature that I think especially new DM's have never even heard of. I don't think "who will search for this" is a sufficient determining factor in removing entries.

> That the monsters are now in categories doesn't change the fact that they won't be used by name anymoreso than they are now.

I disagree. The mere state of being organized will help in the following way:

1. **It necessarily places these lesser-known monsters in the context of similar creatures.** I may not know what "Circe" is, but if it is placed in the category of "Gods/Goddesses -> Greek Gods", that may very well be useful for someone looking to run a Greek-themed campaign. The DM might not search "Circe" out in the search bar, but he may stumble upon it next to "Zeus, Poseidon, Hades," and say, "wait, we've already done all of these. I need a new God...." You never know.

Additionally, categorization accomplishes a few other things:

2. **It exposes duplicates.** Ok, so say nobody can quite agree if "Guardian Goblin Zombie" is too specific or not to get rid of, even if it is only used twice. But we go through and categorize things, run a database query on all categories with a large number of single-use monsters, and see that "zombies" has 80 different monsters with single use, and 15 of them are some variant of "goblin zombie". Good. Now we can officially decide that those can all be condensed down into one or two monsters.

3. **It reveals to Adventure Lookup users which similar monsters already exist.** Say I add a new adventure and start typing in "Super Guardian Goblin Zombie", and then am asked to categorize it. I go through the list and see that there is already a "Goblin Zombie" monster. Adventure Lookup then prompts me something like "ARE YOU SURE THIS NEW MONSTER IS SUFFICIENTLY UNIQUE?". Even if I decide to keep it, a curator can easily approve or disapprove of any new monster that shows up.

4. **It de-clutters the search panel.** I would much rather have 20 or so categories that I can open up and branch down into their own sub-categories *as needed*, than try to wade through a single alphabetical list of 1,000 monsters.

4. **It allows hiding super-unique variant monsters.** Say that *still* nobody can agree to remove the single-use monsters. Perhaps this is a case where "Super Guardian Goblin Zombie" or the named villain "Giblek the Zombie Goblin King" only appears when looking at the actual adventure. We still classify both as a "Goblin Zombie", so those adventures show up in search results, but "Giblek" doesn't need to clutter up the search panel.

**Comment by burnhamrobertp** [comment]
> It necessarily places these lesser-known monsters in the context of similar creatures. I may not know...

This is adventure lookup, not monster lookup. Requiring **everyone** to do this very specific organization of monsters just on the off chance that maybe it could be useful at some time in the future is the same argument that resulted in 3000+ distinct monsters in the monsters list. "Just put all the monsters in, you never know what anyone might want to search for." Now we have a monsters list that's just unmanageable. 

> It exposes duplicates

This is accomplished without categorization, and if you were doing cleanup, the method you're describing **a)** expects that users will categorize things correctly, which is a foregone conclusion that they wont and **b)** can be done by merely looking at all the low-use monsters and using elastic search to find similars.

> It reveals to Adventure Lookup users which similar monsters already exist. 

This is a good sentiment, but the process you're describing would require that - during categorization, you are also going to show them all of the specific monsters that are already inside that category and they're going to see "Goblin Zombie"? Or are you implying that "Goblin Zombie" is a category, which I would flatly disagree with. In any case, the same thing is accomplished by elastic searching the monster's name when the user is typing it in...similar to stack overflow when you go to ask a question..."AdventureLookup deals in general monsters, and I found these similar ones..."

> It de-clutters the search panel

Fewer monsters would also help declutter, which is what I've advocated for in posts above. There are plans to put a text box at the top of several of the main filters, including common and boss monsters, allowing you to type....you'll be able to filter your filter! Yay! :|

> It allows the possibility hiding super-unique variant monsters.

If it doesn't show up, the people who do not agree to remove obviously-unique single-use monsters aren't going to say "oh well its gone, I can't see it anywhere, but you tell me that its not really gone? oh ok that's good enough". When those people advocate for not removing monsters, its not that they want it to be stored somewhere but don't care if it shows up. They are advocating for those monsters to actually show up, so any solution that uses "hiding" them as a compromise won't work. It also doesn't solve the problem of "there are 3000 monsters in the database and that has nothing to do with the use-case of AdventureLookup"

**Comment by calculuschild** [comment]
@burnhamrobertp 

I appreciate that you want to trim a lot of useless monsters out of the list. I am not opposed to that. I think it is important and necessary. My post was not that we need to keep all of the monsters. What I am trying to get across is that *deleting monsters is not enough on its own*.

That said, I am confused as to why you do seem so opposed to categorization. You list a lot of reasons why we need to cut out single-use monsters, which is great. But I still haven't really seen any reasons why categorization should be avoided, other than "sometimes monsters might fit in two categories" which isn't really an issue at all. It really seems like you have some important reason that we shouldn't categorize monsters though, so I think it would be helpful to know what that is.

>  [AdventureLookup] exists to facilitate the finding of adventures...

This is precisely why we need categorization. As a DM, a search bar at the top of the main filters is great, but if I don't know what a monster is called, then I am still no closer to finding the adventure I want. If I can search by category, however, I can at least narrow down my search toward a certain monster type that fits what I am looking for.

On a separate note, I presume that as single-use monsters are being removed, they are at least replaced with their generic version. If an adventure had "Super Guardian Goblin Zombies" as its only monster, the adventure isn't just empty after it's cleaned, right? It at least has "Goblin Zombie" left behind?" We want to refine the information, not leave behind gaps.

**Comment by calculuschild** [comment]
I think it will also be important to devise some standard rules for which monsters should be excised or not. Especially since this seems like a multi-person project. We already have a few:

- **Duplicate monsters via typo**
- **Duplicate monsters via plurality**
- **Monsters with no uses at all** (presumably because they were entered mistakenly or because they were entered but then were cleaned up by a curator at some point)

What else would be necessary? I'm not sure, but I think we can come up with some more guidelines.

As for "Overly specific monsters (human warrior, human rogue, etc.)", I know I have personally searched for adventures featuring rogues or wizards. I have also searched for adventures featuring elves or dwarves. My buddy DM also showed me how he found some good "Druid" adventures for a one-off catering to his Druid, as well as some adventures featuring "half-orcs" catering to his half-orc player. I agree that "Elven Wizard, Elven Rogue, Drow Wizard, Drow Rogue, Human Wizard, Human Rogue" becomes redundant, but I know for a fact that people do search for NPC classes and NPC races. Then we are left with needing to add both "Elf" and "Wizard" as separate monsters in the adventure for the Elven Wizard. Thoughts on this?

On the other hand, I think there are some rules that do not always qualify and we should be careful of:

- **Few-use monsters** (Vecna only shows up in 2 adventures, but is a hugely important villain in D&D lore. I'm certain people will search for him)
- **Monsters "nobody will search for"** (How often is somebody going to search "Otyugh" if they've never heard the word before or even know how to spell it. Yet, it is a legitimate monster in the Monster Manual and shows up in multiple adventures).

**Comment by gn000we** [comment]
I am currently playing with some ideas in my head and hoping to demonstrate them in my dev instance, but i am planning the following (whether i'll actually have time to do it, is another factor):
1) migrate monsters, types etc. Into a tagging structure. This gives complete flexibility.
2) if a tag has more than ten adventures associated with it, it upgrades status to a category.
3) if a categorised set of tagged adventures get tagged with another tag (but same) and is > 5 then the tags are upgraded to sub-category.

No management required.
Q: what if people dont like this system?
A: this is a tagging, backend, change. Front end has the ability to implement it however it wants e.g. as a tree, as a list, as piles, or big mush of tagged adventures refined searches.

4) to advance tagging, the tags can be tagged. This adds a level of meta/semantics to the tags, so "zombie lich ent wizard half orc" can be sufficiently handled.

**Comment by gn000we** [comment]
Additional Q: "yeh, but who handles this?" 
A: there are two roads to take; 1) a public vote consensus (ok, but can cause dissensions) 2) personalised, i.e. each user can define their own, or edit the defaults.

**Comment by gn000we** [comment]
Edit the defaults: personalisation based on profile.
